python/cudf/cudf/core/udf/pipeline.py

```diff
@@ -258,35 +344,25 @@ def compile_or_get(df, f):
     """
 
     # check to see if we already compiled this function
-    frame_dtypes = tuple(col.dtype for col in df._data.values())
-    cache_key = (
-        *cudautils.make_cache_key(f, frame_dtypes),
-        *(col.mask is None for col in df._data.values()),
-        *df._data.keys(),
-    )
+    cache_key = md.cache_key
     if precompiled.get(cache_key) is not None:
         kernel, scalar_return_type = precompiled[cache_key]
         return kernel, scalar_return_type
 
     # precompile the user udf to get the right return type.
     # could be a MaskedType or a scalar type.
-    numba_return_type = get_udf_return_type(f, df)
-
-    _is_scalar_return = not isinstance(numba_return_type, MaskedType)
-    scalar_return_type = (
-        numba_return_type
-        if _is_scalar_return
-        else numba_return_type.value_type
-    )
+    scalar_return_type = get_udf_return_type(md)
+    _check_return_type(scalar_return_type)
+    _is_scalar_return = not isinstance(scalar_return_type, MaskedType)
 
     # this is the signature for the final full kernel compilation
-    sig = construct_signature(df, scalar_return_type)
+    sig = construct_signature(md, scalar_return_type)
 
     # this row type is used within the kernel to pack up the column and
     # mask data into the dict like data structure the user udf expects
-    row_type = get_frame_row_type(df)
+    row_type = get_frame_row_type(np.dtype(list(md.supported_dtypes.items())))
```

Comment:
I had to split this up and play with it in a shell to understand it. Could we try this for readability?
```suggestion
    np_field_types = np.dtype(list(md.supported_dtypes.items()))
    row_type = get_frame_row_type(np_field_types)
```